### ISSUE TITLE
Bug: PointerName rule wrong issue when type name startswith T without prefix

### ISSUE THREAD

**Issue by denieu** [integrated-application-development/sonar-delphi#238]
### Prerequisites

- [X] This bug is in SonarDelphi, not SonarQube or my Delphi code.
- [X] This bug has not already been reported.

### SonarDelphi version

1.5.0

### SonarQube version

10.4.1.88267

### Issue description

The PointerName rule checks whether pointers follow the naming convention.
According to the rule "Why is this an issue?" example:
- ^TObject should be named PObject
- ^TMyType should be named PMyType

Using the T prefix for type/class/record names is good practice, but it is something that will not always be followed.
Rules that validate type/class/record naming can have a prefix setting other than T or can be disabled.

Based on the context above, if a type is defined without the T prefix, but the type name starts with T, the PointerName rule report will be invalid.

For example, a pointer called `PTaskInfo` for the custom type `TaskInfo`, which does not have the prefix T, but its name starts with T, will have an invalid issue reported by the PointerName rule.

I believe the problem can be fixed by changing the check done [here](https://github.com/integrated-application-development/sonar-delphi/blob/master/delphi-checks/src/main/java/au/com/integradev/delphi/checks/PointerNameCheck.java#L58), to something like:
```Java
private static String expectedPointerName(String dereferencedName) {
    return "P" + dereferencedName.substring(dereferencedName.startsWith("T") && 
            dereferencedName.length() > 1 && 
            Character.isUpperCase(dereferencedName.charAt(1)) ? 1 : 0);
}
```

Is my thought valid? Does the proposed fix make sense? In this case, I would like to open a PR with the correction, I'm happy to receive your feedback and suggestions.

### Steps to reproduce

Run Sonar Delphi with the code exemplified in this issue.

### Minimal Delphi code exhibiting the issue

```Pascal
unit MyTestUnit;

interface

type
  TaskInfo = record
    title    : string;
    assignee : string;
  end;
  PTaskInfo = ^TaskInfo;

implementation
begin
end.
```

**Comment by fourls** [comment]
Thanks for raising this issue, @denieu! It's a good point. 

I wonder if we should even raise this issue at all if the originating type is non-compliant with the naming convention. For example, with the type `TaskInfo`, all we know is that the naming convention is broken in some way - we can't assume it is `<no prefix> + TaskInfo` rather than `T + askInfo`, because both equally violate the naming convention.

In my mind there's three possible options we can take in cases where the naming convention is broken like this:

1. allow both `PTaskInfo` and `PaskInfo`, or
2. allow any name as long as it starts with `P`, or
3. allow any name.

My preference is option 1, we don't want to be too permissive as it will reduce the usefulness of the rule. Interested to hear others' opinions.

> I believe the problem can be fixed by changing the check done [here](https://github.com/integrated-application-development/sonar-delphi/blob/master/delphi-checks/src/main/java/au/com/integradev/delphi/checks/PointerNameCheck.java#L58), to something like:
> 
> ```java
> private static String expectedPointerName(String dereferencedName) {
>     return "P" + dereferencedName.substring(dereferencedName.startsWith("T") && 
>             dereferencedName.length() > 1 && 
>             Character.isUpperCase(dereferencedName.charAt(1)) ? 1 : 0);
> }
> ```
> 
> Is my thought valid? Does the proposed fix make sense? In this case, I would like to open a PR with the correction, I'm happy to receive your feedback and suggestions.

Once we've come to a decision on the course of action, a PR would be great! Because the change will probably end up looking pretty similar to this, I have a couple of initial comments:

- The ternary's now a little long and confusing, so it's probably best to restructure this function for added readability
- I'd recommend having a look at `NameConventionUtils.compliesWithPrefix` - its implementation is roughly equivalent to the ternary condition

**Comment by denieu** [comment]
Great considerations, I hadn't thought that `T + askInfo` would be valid based on my argument.

Bringing an extra point for discussion, the [ClassName](https://github.com/integrated-application-development/sonar-delphi/blob/master/delphi-checks/src/main/java/au/com/integradev/delphi/checks/ClassNameCheck.java) rule also allows the E prefix for classes.

Today a class called `EMyClass` conforms to the standard plugin convention. A pointer to this class, in my understanding, should be called `PMyClass`, but the way the PointerName rule was created, the expected name would be PEMyClass.

Should the decision we choose to adjust the T prefix also be applied to E? Should we create a RuleProperty in the PointerName rule with valid prefixes for the types?

**Comment by fourls** [comment]
That sounds like a good idea - this would bring PointerName in line with HelperName's behaviour, which makes sense (both enforce a naming convention for types that are logically derived from another type).

Here's HelperName's properties:

https://github.com/integrated-application-development/sonar-delphi/blob/a65945b69fca45b84bd49f75dc6713d95dc27779/delphi-checks/src/main/java/au/com/integradev/delphi/checks/HelperNameCheck.java#L48-L58

PointerName could have basically the same properties.

**Comment by fourls** [comment]
Thinking about it, a rule property approach has some issues - for example, allowing both `T` and `Tprefix` would allow `TprefixFoo`'s pointer type to be `PFoo`, which is wrong.

I don't think it's necessary for us to make the rule configurable - it might be better to do a variant of my original suggestion and enforce that a pointer type is the result of:

1. (optionally) removing an initial `T`, `E`, or `I` 
2. adding a `P`

This should allow all acceptable pointer names while keeping false negatives to a minimum.

**Comment by denieu** [comment]
> Thinking about it, a rule property approach has some issues - for example, allowing both `T` and `Tprefix` would allow `TprefixFoo`'s pointer type to be `PFoo`, which is wrong.
I'm not sure if I understand the problem correctly, but I believe this problem can be avoided in a configurable implementation of the rule.

Reflecting on the best way to improve the PointerName rule, I believe that making the rule configurable and slightly changing the way it works is the best option.

My suggestion is the following:
1. Add a RuleProperty called `pointerPrefixes`, similar to the HelperName rule
2. Add a RuleProperty called `extendedTypePrefixes`, similar to the HelperName rule
3. If the type referenced by the pointer is `NameConventionUtils.compliesWithPrefix` with some `extendedTypePrefixes` prefix, the pointer name must be the same as the name of the referenced type but with one of the `pointerPrefixes` prefixes instead of the current prefix
4. If the type referenced by the pointer is in PascalCase, but does not have any of the prefixes stored in `extendedTypePrefixes`, the pointer name must be one of the `pointerPrefixes` prefixes + the current name of the type
5. If the type referenced by the pointer is not in PascalCase or has a prefix that is not present in `extendedTypePrefixes`, the name of the pointer only needs to be `NameConventionUtils.compliesWithPrefix` with any prefix present in `pointerPrefixes`, no reference to the name original of the referenced type needs to be made

The valid names for each type can be seen in the table below (consider the default values similar to the HelperName rule):
| Referenced Name | Pointer Name |
|--------|--------|
| TMyType | PMyType |
| EMyType | PMyType |
| MyType | PMyType |
| TaskInfo | PTaskInfo |
| EvenNumber | PEvenNumber |
| invalid_type_name | PValidPointerName |

The main changes with this suggestion are:
- The PointerName rule is self-contained, that is, if the original name of the type is invalid and the pointer is valid, no issue will be added
- The PointerName rule is configurable and therefore suitable for more contexts

If any point in the explanation was not clear, I am happy to clarify any doubts. 
If what I proposed makes sense, I can open a PR for the changes described, I believe it will be simpler to understand the proposal with an example of code available, even if it needs to be redone if another solution is chosen.

**Comment by fourls** [comment]
While this would make the rule very flexible, it's also open to more false negatives and requires the user to have configured the rule correctly.

> 4. If the type referenced by the pointer is in PascalCase, but does not have any of the prefixes stored in `extendedTypePrefixes`, the pointer name must be one of the `pointerPrefixes` prefixes + the current name of the type

Even if a codebase has a prefix convention, the user still want to support adding pointer types for the standard library and 3rd parties without the prefix. This means that the `extendedTypePrefixes` property would always contain `T`, and the `pointerPrefixes` property would always contain `P`.

In an example codebase with an `xy` prefix convention, `extendedTypePrefixes` would be `T,Txy,Exy,Ixy` and `pointerPrefixes` would contain `P,Pxy`, which is undesirably permissive - for example:

* `TxyMyClass` to `PMyClass` is OK, even though it shouldn't be (the correct form is `PxyMyClass`)
* `TStringList` to `PxyStringList` is OK, even though it shouldn't be (the correct form is `PStringList`)

This is what I was trying to say in my last comment. What we actually want is a way to enforce a mapping from the extended type prefix to the corresponding pointer prefix, which would give us something like this:

* `T`, `E`, `I` to `P`
* `Txy`, `Exy`, `Ixy` to `Pxy`

Which is a long way of saying that

> a pointer type is the result of:
> 1. (optionally) removing an initial T, E, or I
>  2. adding a P

Which means that we can enforce the same behaviour by checking these rules, instead of making the user do unnecessary configuration. The only thing that I think this doesn't support is custom prefixes for types that don't follow the naming convention - I suppose we could deal with this by either
1. Having the `pointerPrefixes` property and changing rule 2 to "adding one of the pointer prefixes" (this has the side effect of allowing `TStringList` to `PxyStringList`, which I'm not super keen on)
2. Not enforcing the rule at all for types that don't follow the naming convention

Ideally, I'd rather keep everything being configurable to a last resort, since it's possible to detect 99% of cases another way, it requires the user to exert non-trivial effort in finding the rule, understanding it, and configuring it, and makes the default quality profile less useful.

> If what I proposed makes sense, I can open a PR for the changes described, I believe it will be simpler to understand the proposal with an example of code available, even if it needs to be redone if another solution is chosen.

Feel free to open a PR whenever you have an implementation you're happy with! Discussion can always continue there in the form of review.

**Comment by denieu** [comment]
Now everything is clearer, your suggestions made a lot of sense to me.

I considered the points raised in this discussion and created this PR https://github.com/integrated-application-development/sonar-delphi/pull/240.

I would appreciate your review @fourls.
